Initial Code

brentru opened the pull request:
CircuitPython_RSA is a port of [Python-RSA, a pure-Python RSA implementation](https://github.com/sybrenstuvel/python-rsa). 

Modifications from initial library. 
* The command-line-specific (utils.py) tools have been removed from the library to save space.
* CPython Logging has been replaced with the Adafruit CircuitPython Logger Module.
* Miller–Rabin Primality test algorithm modified to support library's `fast_pow`, CircuitPython's `pow()` does not support modular exponentiation.
* `adafruit_rsa/tools` added, pure-python `binascii`, `warnings`, and `hashlib` modules.

**Support for:**
* RSA Key pair (private/public) generation
* Message encryption and decryption
* Signature creation and verification
  * w/SHA-256, SHA-384, SHA-512

Examples are provided for signature creation/verification, keypair generation and encryption/decryption, and a unit-test-like for testing future modifications to this library.

Commit:
cookiecut!

Commit:
add diff'd files

Commit:
rename import statements to match file directory

Commit:
add /tools directory, add a hashlib implementation in pure-python. NOTE: this implementation does not have SHA1

Commit:
remove micropython references to uhashlib, replace with hashlib, remove sha1 reference, add mit license to top of init

Commit:
add hexlify definition

Commit:
add working hexlify, cleanup b2a/a2b methods to reflect latest micropython-lib master

Commit:
rsa-python doesnt fail when imported anymore, modification made to do modular exponents

Commit:
remove doctest, privatekey repr works

Commit:
fixup readme to credit original impl.

Commit:
prune parallel, cli, and pkcs1_v2 modules from library for size-saving

Commit:
fix raw_bytes in int2bytes

Commit:
re-enable supported SHA224/384/512 hashlib methods, disable unsupported MD5 and sha1 hashes

Commit:
mock out classes for MD5 and SHA1

Commit:
rollback locally, fix sha256

Commit:
add a logging kwarg to newkey generation

Commit:
add mock-unittest-like test for this library

Commit:
add pyasn1

Commit:
add modified import for tools.pyasn1 dir

Commit:
remove pyasn1 cruft

Commit:
remove pyasn1-required PEM file loading, wont work with this library so far until a pyasn1 lib exists for circuitpython

Commit:
add string translate method

Commit:
add jwt generation script

Commit:
pylint init, compat

Commit:
pylint asn1, common, key

Commit:
linting..

Commit:
remove util

Commit:
bringin from board

Commit:
linttest init, compat

Commit:
linted rsa/ directory and refactored

Commit:
lint/refactor binascii

Commit:
remove string.py, not required by RSA

Commit:
make warnings subclass impl descriptive

Commit:
refactor and lint hashlib

Commit:
add docus

Commit:
update examples

Commit:
pylint

Commit:
check in the rsa/ dir instead

Commit:
add RSA test

Commit:
lint

Commit:
lint again

Commit:
fix python bad option value throwing the same pylint

Commit:
remove top-level tools folder

ladyada reviewed (commented):
hmm dya want to split hashlib out? i think that would be useful separate rather than embedded?

brentru commented:
@ladyada I could split hashlib out - I held off splitting it off since it'll eventually be a core module. Should I also bring `binascii` out as well so we have pure-python versions of these two modules?  The`JWT` module, and prob. a lot of future modules currently depends on the `hashlib` and `binascii` from this library, for example.

Commit:
removing tools, deprecate warnings in favor of logging instead

Commit:
replace with adafruit_circuitpython_ repos, only import what's required

Commit:
correct path to hashlib.md5/hashlib.sha1

Commit:
linting

Commit:
update configuration for sphinx

Commit:
credit in license

brentru commented:
@ladyada Removed the `adafruit_rsa/tools/` folder. [Hashlib](https://github.com/adafruit/Adafruit_CircuitPython_hashlib) and [binascii](https://github.com/adafruit/Adafruit_CircuitPython_binascii) have been split out into new repositories. Adafruit_RSA has been modified to reflect these changes.

Replaced `tools/warnings.py` with the circuitpython logger module, instead of warnings.

Updated code successfully passes tests
> Press any key to enter the REPL. Use CTRL-D to reload.
> Adafruit CircuitPython 4.1.0 on 2019-08-02; Adafruit PyPortal with samd51j20
> >>>
> soft reboot
> Auto-reload is on. Simply save files over USB to run them or enter REPL to disable.
> code.py output:
> Testing: <function test_encrypt_decrypt at 0x20003400>
> 2314.25: DEBUG - find_p_q(128): Finding p
> 2316.53: DEBUG - find_p_q(128): Finding q
> OK!
> Testing: <function test_mod_msg at 0x200033e0>
> 2317.87: DEBUG - find_p_q(128): Finding p
> 2318.51: DEBUG - find_p_q(128): Finding q
> OK!
> Testing: <function test_randomness at 0x200033f0>
> 2322.82: DEBUG - find_p_q(128): Finding p
> 2325.37: DEBUG - find_p_q(128): Finding q
> OK!
> Testing: <function test_sign_verify_sha256 at 0x20003410>
> 2329.66: DEBUG - find_p_q(248): Finding p
> 2356.35: DEBUG - find_p_q(248): Finding q
> OK!
> Testing: <function test_sign_verify_sha384 at 0x20003460>
> 2376.34: DEBUG - find_p_q(312): Finding p
> 2449.32: DEBUG - find_p_q(312): Finding q
> OK!
> Testing: <function test_sign_verify_sha512 at 0x20003510>
> 2530.8: DEBUG - find_p_q(376): Finding p
> 2586.87: DEBUG - find_p_q(376): Finding q
> OK!
> Ran 6 tests in 283.514 seconds
>
